### PR TITLE
Update tonal and create addChord function for custom chords

### PR DESCRIPTION
**Pull request by mickmister**
Using the latest version of `tonal`, we can create custom chord shapes to use in scribbletune clips. The tonal package uses its own internal state to store all of its chords. You can add new chords using [`Tonal.ChordType.add`](https://github.com/tonaljs/tonal/blob/4af3763c34780d04b577f40367e267423a6e11dc/packages/chord-type/index.ts#L93). This PR exports this function from the scribbletune package through a new function `addChord`.

I was thinking of exporting all of the `tonal` package, but that is pre-optimization for the sake of this PR. I'm not sure what other useful things would be appropriate to expose from `tonal`.

All unit tests are passing, though I don't know how to do thorough regression testing for this change. The project somehow doesn't have types for `Array.prototype.flat`, so I temporarily polyfilled the function in `utils.ts` to make the project compile.

Fixes https://github.com/scribbletune/scribbletune/issues/110

### CODE REVIEW

File: examples/chords.js [hunk 1]
```diff
@@ -13,7 +17,7 @@ scribble.midi(clip);
 
 const clip2 = scribble.clip({
   notes: scribble.getChordsByProgression('C4 major', 'vi IV V I'),
-  pattern: 'x',
+  pattern: 'xxxx',
```
Comment (mickmister) [inline examples/chords.js:20]: Without changing this, it only had the `vi` chord in the output. I tested on master before my changes as well, and it had the same unwanted behavior.

**Comment by mickmister** [comment]
There is still an opportunity for adding shorthand chord creation in the built-in scribbletune chord notation, but I thought this was a good first step in allowing both contributors and application developers using the library to use the custom chord registration exposed by tonal.

Since the application developer could want all sorts of chord shapes, maybe we post some examples of some practical chord shapes without necessarily cluttering the chord type namespace with the library itself? There are a lot of options, and this PR seemed like the most flexible, backwards-compatible step right now.

cc @qchenevier so you're aware of this PR. I've been thinking about the "smooth chord transition" idea. I really think that should be up to the application developer as well, but we can have some functionality that makes this easier. Maybe a "for the second chord of this progression, have scribbletune pick one of these chords on the fly", some of which would happen to be closer to the previous one, and the library could then approximate which one is closest in proximity to the previous one. That is a bit opinionated though. Once again, lots of options here.

I chose to continue discussion here as I think it makes sense for this small-scope topic to be public.

File: examples/chords.js (hunk 1 above)
Comment (walmik) [reply examples/chords.js:17]: Yes the notes repeat if the pattern is longer but the other way round is not yet implemented.

File: tests/chord.spec.ts [hunk 2]
```diff
@@ -34,4 +34,12 @@ describe('../src/chord', () => {
     expect(getChord('CM-5')!.join()).toBe('C5,E5,G5');
     expect(getChord('Gbm')!.join()).toBe('Gb4,A4,Db5');
   });
+
+  it('returns a custom chord', () => {
+    addChord(['1P', '5P', '8P', '22P', '24M', '26P'], ['Q1', 'Q2'], 'full');
```
Comment (walmik) [inline tests/chord.spec.ts:39]: This is so cool!

**Comment by walmik** [comment]
@qchenevier if you get a moment, can you please take a look as well?

**Comment by qchenevier** [comment]
@walmik Yeah sure. I was willing to do it, but I have been a bit overwhelmed lately. 😅 Tomorrow morning (western europe time), I'll have some free time for this.

**Comment by mickmister** [comment]
@walmik Glad you like the feature!

For a followup feature, I have a `composeChord` function to allow you to combine pre-existing and adhoc chords into a larger chord without having to type the sort of cryptic values of `['22P', '24M', '26P']` etc. The resulting chord takes into account the root octave, appending the `-(octave)` of the lowest octave for you.

I'm curious what you think of this API as I'm working on it:

```js
scribble.addChord(['1P', '5P', '8P'], [], 'power');
scribble.addChord(['1P', '3M', '5P', '8P'], [], 'triad-octave');

const chord = scribble.composeChord(['cc'], 'coolchord', [
  {octave: 2, chord: 'power'},
  {octave: 4, chord: 'triad-octave'},
  {octave: 5, intervals: ['3M', '5P']},
]);

const clip = scribble.clip({
  notes: ['A' + chord.shorthand, 'G' + chord.shorthand],
  pattern: 'xx',
});

console.log(chord);
/*
{
    rootOctave: 2,
    chord: fullChordFromTonal,
    shorthand: 'cc-2',
}
*/
```

File: src/utils.ts [hunk 3]
```diff
@@ -80,3 +80,8 @@ export const pickOne = (arr: any[]) =>
  * Boolean generator
  */
 export const dice = () => !!Math.round(Math.random());
+
+/**
+ * Polyfill for Array.prototype.flat
+ */
+export const flat = (arr: any[][]) => arr.reduce((acc, val) => acc.concat(val), []);
```
Comment (mickmister) [inline src/utils.ts:87]: @walmik I personally don't think this function belongs in the project. Do you know what needs to be done for this function to be properly available without the polyfill?

**Comment by walmik** [comment]
Sorry @mickmister ! Looks like the BPM PR #112 merge has conflicted index.js.. you might need to npm run build once. Will merge right after.

**Comment by qchenevier** [comment]
> Sorry @mickmister ! Looks like the BPM PR #112 merge has conflicted index.js.. you might need to npm run build once. Will merge right after.

@walmik About that, what do you think of having the build done in the CI, rather than commiting the locally-built files ?

**Comment by walmik** [comment]
That s a good idea @qchenevier ! We do have a travis.yml, is it just a matter of adding the `npm run build` command [here](https://github.com/scribbletune/scribbletune/blob/master/.travis.yml#L7)? I m curious to know whether it will also create a commit for the file(s) built. Coz we need the transpiled file(s) _committed_ [for npm publish as well as CDNjs autoupdate which kicks in for npm publish]

**Comment by mickmister** [comment]
> Coz we need the transpiled file(s) committed [for npm publish as well as CDNjs autoupdate which kicks in for npm publish]

How often is the CDNjs updated? Is it just when a new release is made? Maybe we could just commit the compiled version in each release bump PR?

**Comment by mickmister** [comment]
Conflicts resolved. There were also some tslint warnings I corrected.

**Comment by walmik** [comment]
@mickmister CDNjs must have some cronjob to go over all the libs they host and that script might just check the current version against the one published on npm. Typically I see a publish make it to CDNjs within a day or two.

**Pull request merged**